### PR TITLE
Remove etcd gateway

### PR DESCRIPTION
**Pull request by Quentin-M**
Hey.

@s-urbaniak could you take a look at the second commit and tell me why this line was the way it was? It seems like you introduced the `127.0.0.1:2379` endpoint in there.

@s-urbaniak @squat From looking on what gets created on AWS, it looks like the security groups regarding to etcd are not too bad? Is there something we need to remove there?

### CODE REVIEW

**Comment by Quentin-M** [comment]
Ref: https://github.com/coreos/tectonic-installer/pull/135/files#diff-69319ec0b8e4196b069913c3780a03a6R37

**Comment by sym3tri** [comment]
see https://github.com/coreos/tectonic-installer/issues/253

**Comment by s-urbaniak** [comment]
@Quentin-M The current logic is:

1. If experimental mode is enabled (self-hosted etcd), then use `var.etcd_service_ip`.
2. Else if **no** etcd TLS certificates are provided, i.e. we need to bootstrap an etcd node ourselves, then use the etcd gateway reachable under `http://127.0.0.1:2379`.
3. Else (if etcd TLS certific _are_ provided), use the user-provided `var.etcd_endpoints`.

**Comment by s-urbaniak** [comment]
@Quentin-M regarding security groups in my opinion we don't have to do anything as removing the gateway retains the necessity to talk via port 2379/2380.

Just as a side-note in the context of self-hosted etcd: we skip the creation of the etcd security group at all, if experimental is enabled: https://github.com/coreos/tectonic-installer/blob/b87a8af/platforms/aws/main.tf#L13

File: modules/bootkube/assets.tf [hunk 1]
```diff
@@ -28,7 +28,7 @@ resource "template_dir" "bootkube" {
     etcd_operator_image    = "${var.container_images["etcd_operator"]}"
     kenc_image             = "${var.container_images["kenc"]}"
 
-    etcd_servers    = "${var.experimental_enabled ? format("http://%s:2379", var.etcd_service_ip) : data.null_data_source.etcd.outputs.no_certs ? "http://127.0.0.1:2379" : join(",", formatlist("https://%s:2379", var.etcd_endpoints))}"
+    etcd_servers    = "${var.experimental_enabled ? format("http://%s:2379", var.etcd_service_ip) : data.null_data_source.etcd.outputs.no_certs ? join(",", formatlist("http://%s:2379", var.etcd_endpoints)) : join(",", formatlist("https://%s:2379", var.etcd_endpoints))}"
```
Comment (sym3tri) [inline modules/bootkube/assets.tf]: That would be a good comment to put in the code

**Comment by Quentin-M** [comment]
Thanks for the feedback. I will address it and test that as soon as I get to the office. Was playing with several things at the same time yesterday but then got a bit confused by the localhost assumption - even though I guessed it'd be for talking to the gateway, it felt weird that we pass the endpoints as a parameter but discard them. Seems a bit more reasonable now - the bootkube module should be able to be used independently of any platform or setup.

File: modules/bootkube/assets.tf (hunk 1 above)
Comment (Quentin-M) [inline modules/bootkube/assets.tf]: Yes, any endpoint computation should be done by the caller. Ideally, the etcd module of each platform should always output the right set of endpoints depending on the situation - it has all the required data. This avoids repeating the logic at any other places and simplify our implementation overall.

Comment (Quentin-M) [inline modules/bootkube/assets.tf]: Will keep it as is now and add a comment.

**Comment by Quentin-M** [comment]
- Tested with provisioned etcd
- Tested with self-hosted etcd
Works in both cases. Good to merge afaik.

**Comment by Quentin-M** [comment]
@sym3tri Jenkins breaks due to reasons that are not that PR.

```
31 error(s) occurred:

* module.vpc.aws_main_route_table_association.main_vpc_routes: Resource 'aws_route_table.default' does not have attribute 'id' for variable 'aws_route_table.default.id'
* module.vpc.aws_route_table_association.route_net[2]: Resource 'aws_route_table.default' does not have attribute 'id' for variable 'aws_route_table.default.id'
* module.vpc.aws_security_group_rule.master_ingress_etcd: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.worker_ingress_heapster_from_master: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_services_from_console: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_icmp: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_flannel_from_worker: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group.etcd: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.worker_ingress_node_exporter_from_master: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_node_exporter_from_worker: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_flannel: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_https: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.worker_ingress_flannel_from_master: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_route_table_association.route_net[1]: Resource 'aws_route_table.default' does not have attribute 'id' for variable 'aws_route_table.default.id'
* module.vpc.aws_security_group_rule.master_ingress_kubelet_insecure_from_worker: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_route_table_association.route_net[0]: Resource 'aws_route_table.default' does not have attribute 'id' for variable 'aws_route_table.default.id'
* module.vpc.aws_security_group_rule.master_ingress_ssh: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.worker_ingress_kubelet_secure_from_master: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_bootstrap_etcd: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_kubelet_secure: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_kubelet_secure_from_worker: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_heapster: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_heapster_from_worker: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_node_exporter: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_http: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_nat_gateway.nat_gw[2]: 1 error(s) occurred:

* aws_nat_gateway.nat_gw.2: Error creating NAT Gateway: InvalidSubnet: Subnet subnet-fac331a1 was not found
	status code: 400, request id: b53c73ae-038c-42f9-9391-08aa5f8431b4
* module.vpc.aws_security_group_rule.master_egress: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.master_ingress_kubelet_insecure: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_route.igw_route: Resource 'aws_route_table.default' does not have attribute 'id' for variable 'aws_route_table.default.id'
* module.vpc.aws_security_group_rule.master_ingress_services: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'
* module.vpc.aws_security_group_rule.worker_ingress_kubelet_insecure_from_master: Resource 'aws_security_group.master' does not have attribute 'id' for variable 'aws_security_group.master.id'

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.[0m[0m
make: *** [apply] Error 1

1 error(s) occurred:

* module.vpc.aws_vpc.new_vpc (destroy): 1 error(s) occurred:

* aws_vpc.new_vpc: DependencyViolation: The vpc 'vpc-b62915d1' has dependencies and cannot be deleted.
	status code: 400, request id: c6dc211f-b5cc-44e2-aed4-d142fc9872e4
```

**Comment by Quentin-M** [comment]
Re-started anyways.

**Comment by Quentin-M** [comment]
Again, unrelated.

```
[31mError applying plan:

1 error(s) occurred:

* module.vpc.aws_nat_gateway.nat_gw[2]: index 2 out of range for list aws_eip.nat_eip.*.id (max 2) in:

${aws_eip.nat_eip.*.id[count.index]}
```

**Comment by sym3tri** [comment]
@Quentin-M unrelated but we have to fix the failing test.

**Comment by Quentin-M** [comment]
Passed this third time.

**Comment by sym3tri** [comment]
ok, i am looking into the test

**Pull request merged**